### ISSUE TITLE
Use tracker API project status to restrict requests for matrix service manifests to only complete projects and warn the user of any excluded projects

### ISSUE THREAD

**Issue by NoopDog** [DataBiosphere/data-browser#838]
Currently on the Request Concatenated Expression Matrix flow its easy to get a "mismatch error" where the browser sends matrix files in the manifest, where the matrix file's bundle has not been loaded into the matrix service. 

Once this happens the Matrix Service errors without returning partial results and the user must try to determine what requested data is causing the error. 

This can happen because there is no current mechanism to keep the Matrix Service and Azul in sync or for Azul to know what data is loaded in to the Matrix Service. While we discuss longer term approaches to solving this problem it is possible we can appear to be synchronized at the project level by using tracker API to indicate if a project is loaded into the  Matrix Service of if it is Not Expected to be loaded into the matrix service.

To use the tracker API we would:

1. Hit the tracker API on page load or "get data" page load and obtain the completion status for all projects. 

1. When the user navigates to the "get data" page we will  check if *ANY* of the projects listed in the term facts under the projects term are "complete" according to the tracker. If projects are complete, we will enable the "Start" button under  "Create Concatenated Expression Matrix" otherwise we will disable it. with a message explaining the supported data types.

1. Once the user presses "Start" on the  Create Concatenated Expression Matrix, we will check if *ALL* of the projects requested by the user are marked complete by the tracker.

 If *ALL* are supported, we will  proceed with the flow and request the expression matrix. 

If only *SOME* projects are supported we will warn and list the criteria for including a project in the request. We will also continue with the flow and request the manifest for preparing the expression matrix, however, we will modify the users query to deselect any non complete projects and select the complete projects that are still available in the project facet after the users other terms have been selected. 

In a future ticket we may modify the summary to show exactly what projects will be returned and show the related counts.

The above flow will synchronize the data portal and matrix service at the project level according to the trackers conception of "complete". Note that:

1. The tracker currently hits the ingest API to enumerate the submission envelopes and cycles through them to update the status of any previously incomplete envelopes and add any new envelopes to its inventory.

1.  Once a submission envelope is complete the tracker can query DSS for the number of primary bundles in the project.

1. The number of secondary analysis bundles and number of bundles loaded in to the matrix service for a project are expected to match the number of primary bundles from the completed submission envelope. 

1. The tracker knows if secondary analysis or matrix service support is expected according to the submission data properties such as species, library construction method (and details such as paired end, fluidigm )

1. Projects are complete when all stages are complete, or all earlier stages are complete and all later stages are "not expected" according to the tracker rules. (to refine) 

1. If a project is modified or added to, this will result in a new submission envelope and the project status will change to incomplete until the new submission envelope is processes.  

1. For projects where only some of the data is expected to be processed (e.g. human and mouse data in the same project) the tracker currently will assume all of the data will be processed so the mixed projects will never go to complete for 2ndary analysis or beyond. There is a ticket to address this in the tracker. 
Until this is resolved, mixed projects would be removed from the users request when requesting a manifest destined for the matrix service. (which may not be a problem as they currently are not in the matrix service any how at the moment at least)
Also, requesting only a mixed project would result in the matrix "Start" button being greyed out with explanation that the data type is not supported.

**Comment by parthshahva** [comment]
So this ticket is focused on the data browser x matrix interaction. Should we also utilize the tracker api to inform users of incomplete/out-of-date analysis or primary data in other interactions (for example if they go to download the manifest for a project)? Or maybe a visual flag that allows them to select the incomplete/out-of-date primary/analysis data but at least be warned of the state

**Comment by NoopDog** [comment]
@hannes-ucsc @theathorn @MillenniumFalconMechanic @parthshahva Does the above sound reasonable? Is this something we want to do while we are working out how to more tightly sync the Matrix Service and Azul?

**Comment by NoopDog** [comment]
@parthshahva incomplete processing indication will be refined here:
https://app.zenhub.com/workspaces/dcp-5ac7bcf9465cb172b77760d9/issues/humancellatlas/data-browser/688

**Comment by hannes-ucsc** [comment]
I don't think we should hold off spending more time on this. The central question is whether it is acceptable or even desirable for the matrix service to expose partial matrices, i.e. matrices that are are missing bundles either temporarily or longer term. My uneducated guess is that yes, this desirable and no, that it is not harmful. However I am not to decide. @calvinnhieu and I are working on a proposal document that lays out the problem as well as the pros and cons of various solutions. If anyone wants to join the effort, they more than are welcome and should let us know. We will share the document with them. Once the document is done, we plan on presenting it to project and scientific leadership for a decision.

**Comment by diekhans** [comment]
In my experience, it is *highly* undesirable to release
incomplete matrices.

RFC Processing Datasets that Span Multiple Data Collection Runs
https://github.com/HumanCellAtlas/dcp-community/pull/88

address this issue as well.

I have no idea if it is less work to move forward with this RFC
once approved or do this tracker approach.

**Comment by NoopDog** [comment]
Yes, the he tracker approach is offered as an interim solution to preventing the data browser from requesting data that the matrix service does not have. This basically synchronizes the data browser and Azul on "Complete Projects" (or complete submission envelopes) and would prevent the mismatch errors while a deeper integration is being designed and built. 

The solution above can be in place in a week or two. The site is GA now and currently many searches will result in the "mismatch  error". Personally, I would like to be able to avoid these kinds of errors fairly quickly. 

One other option is to have the matrix team just serve the partial matrix, and warn the browser that its incomplete, but my understanding is that @mckinsel would rather not return partial, incomplete matrices to the user, and that its too difficult to tell exactly what was missing from the matrix. 

I agree, that we need to escalate the question about if partial matrix results are useful.

Also I was reminded that while Azul is currently processing bundle requests in real time @jahilton is starting to talk about letting users QC their (analyzed) data in the system before it is available for everyone to see. 

If we are moving to a setup where secondary analysis is complete and has had some level of QA/QC done, before we show the show project at all, this may have some ramifications for the Azul/Matrix service integration flow.

I appreciate that we want to get the data out to users as quickly as possible so they can start to analyze and let us know how it looks etc so we can improve it if required. This is a very strong point raised by @hannes-ucsc in my opinion. @mckinsel did point out, however ,that the time between a project going from incomplete to complete might only really need to be a few hours or a few days. 

Once the processing pipelines are more reliable, and especially if we take data similar to data already processed, the time to process a project should come down and may lessen the benefit of synchronizing the systems, cell by cell, or bundle by bundle, and might make the synchronizing by submission envelope a reasonable thing to do.

**Comment by theathorn** [comment]
I agree with Dave's 1st paragraph in the previous comment. It's a better UX than getting errors from the MS request.

**Comment by theathorn** [comment]
8/29 meeting decision is to use the Matrix Service API v1 to determine if a project is matrix-able.

**Comment by NoopDog** [comment]
@theathorn can  we close this then and create a new ticket to represent researching the updated approach?

The data browser can query  the matrix service to determine what projects they have, but we can not tell if the project is complete in the matrix service or if azul and the matrix service have the same bundles.

One approach we discussed is that Azul adds to its bundle index and indicator that shows if the Matrix service has indexed a given bundle and uses that index to restrict the bundles sent in the manifest to the matrix service to only bundles that the matrix service has.

The goal here is to stop gettin the "Missing bundles error"  when attempting to create a matrix from a browser query.

**Comment by theathorn** [comment]
I talked with @hannes-ucsc and he will work with @NoopDog  on updating the proposed Matrix Service interaction in https://docs.google.com/document/d/1o6zIGHmzVWulM-Lo2dk4pXjC_JqcixOx1IJ1g3mbTIY/.

The goal is to interact with the MS to determine which projects have matrices available and to further address the issue of "partial" support where the MS has a matrix for some but not all of the data in a project (e.g. MS has "human" but not "mouse" data).

Moving this ticket to the Product Backlog until we have an agreed plan.

**Comment by NoopDog** [comment]
Closing, tracker API and Matrix API will be sunset
